intern/cycles/kernel/kernel_color.h

```diff
@@ -33,6 +33,673 @@ ccl_device float linear_rgb_to_gray(KernelGlobals *kg, float3 c)
+        make_float4(815.0, 0.000003581652f, 0.000001293400f, 0.000000000000f),
+        make_float4(816.0, 0.000003339127f, 0.000001205820f, 0.000000000000f),
+        make_float4(817.0, 0.000003112949f, 0.000001124143f, 0.000000000000f),
+        make_float4(818.0, 0.000002902121f, 0.000001048009f, 0.000000000000f),
+        make_float4(819.0, 0.000002705645f, 0.000000977058f, 0.000000000000f),
+        make_float4(820.0, 0.000002522525f, 0.000000910930f, 0.000000000000f),
+        make_float4(821.0, 0.000002351726f, 0.000000849251f, 0.000000000000f),
+        make_float4(822.0, 0.000002192415f, 0.000000791721f, 0.000000000000f),
+        make_float4(823.0, 0.000002043902f, 0.000000738090f, 0.000000000000f),
+        make_float4(824.0, 0.000001905497f, 0.000000688110f, 0.000000000000f),
+        make_float4(825.0, 0.000001776509f, 0.000000641530f, 0.000000000000f),
+        make_float4(826.0, 0.000001656215f, 0.000000598090f, 0.000000000000f),
+        make_float4(827.0, 0.000001544022f, 0.000000557575f, 0.000000000000f),
+        make_float4(828.0, 0.000001439440f, 0.000000519808f, 0.000000000000f),
+        make_float4(829.0, 0.000001341977f, 0.000000484612f, 0.000000000000f),
+        make_float4(830.0, 0.000001251141f, 0.000000451810f, 0.000000000000f)
+    };
+    return find_position_in_lookup_unit_step(wavelength_xyz_lookup, wavelength, 360, 830);
+}
+
+ccl_device float3 wavelength_intensities_to_linear(KernelGlobals *kg, float3 intensities, float3 wavelengths)
+{
+    float3 sum_x_xyz = wavelength_to_xyz(wavelengths.x) * intensities.x;
+    float3 sum_y_xyz = wavelength_to_xyz(wavelengths.y) * intensities.y;
+    float3 sum_z_xyz = wavelength_to_xyz(wavelengths.z) * intensities.z;
+    float3 xyz_sum = sum_x_xyz + sum_y_xyz + sum_z_xyz;
+
+    return xyz_to_rgb(kg, xyz_sum);
+}
+
+ccl_device float3 find_position_in_lookup(float4 lookup[], float wavelength)
+{
+    int upper_bound = 0;
+    while(lookup[upper_bound][0] < wavelength) {
+        upper_bound++;
+    }
+    int lower_bound = upper_bound -1;
+    float progress = (wavelength - lookup[lower_bound][0]) / (lookup[upper_bound][0] - lookup[lower_bound][0]);
+    return make_float3(
+        float_lerp(lookup[lower_bound][1], lookup[upper_bound][1], progress),
+        float_lerp(lookup[lower_bound][2], lookup[upper_bound][2], progress),
+        float_lerp(lookup[lower_bound][3], lookup[upper_bound][3], progress)
+    );
+}
+
+ccl_device float find_position_in_lookup_2d(float2 lookup[], float wavelength)
+{
+    int upper_bound = 0;
+    while(lookup[upper_bound][0] < wavelength) {
+        upper_bound++;
+    }
+    int lower_bound = upper_bound -1;
+    float progress = (wavelength - lookup[lower_bound][0]) / (lookup[upper_bound][0] - lookup[lower_bound][0]);
+    return float_lerp(lookup[lower_bound][1], lookup[upper_bound][1], progress);
+}
+
+ccl_device float3 linear_to_wavelength_intensities(float3 rgb, float3 wavelengths)
```

Comment:
Here's some stuff that I would ideally like to replace in the future, but it is what is allowing us to use RGB data in a spectral renderer. Some way or another, we need to take an RGB value (linear sRGB in this case, A.K.A. REC.709) and convert it into a spectrum.

There's lots of information lost when getting an RGB value from a spectrum, so we're never going to be able to do that reconstruction perfectly, but this is one method of getting a usable spectrum which fits a bunch of criteria, namely:
1. The XYZ coordinate of the reconstructed spectrum must be identical to the XYZ coordinate of the input colour
2. Spectra can never exceed 1 or go below 0, so that they retain physical meaning as a 'reflection spectrum'.
3. RGB 1,1,1 should equal a flat spectrum of 1.
4. The spectrum should have the least slope possible.

With these constraints, you get a nice average (but boring) spectrum you can use in place of an RGB value.

I'm hoping in to future to replace this with another method by Wenzel Jakob and Johannes Hanika outlined in the paper [A Low-Dimensional Function Space for Efficient Spectral Upsampling](https://rgl.s3.eu-central-1.amazonaws.com/media/papers/Jakob2019Spectral_3.pdf) which uses a really clever representation of a spectrum, but requires a bit of preprocessing to use. Benefits are that it works for many more colours than sRGB (wider gamut) and might actually be _faster_ at runtime. It'll be quite a lot of work to implement since it needs some significant changes to how Cycles works but it's one of the goals.